bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/internal/VoiceManagerImpl.java

```diff
@@ -637,11 +655,24 @@ public HumanLanguageInterpreter getHLI(String id) {
 
     @Override
     public Set<Voice> getAllVoices() {
-        Set<Voice> voices = new HashSet<>();
-        for (TTSService tts : ttsServices.values()) {
-            voices.addAll(tts.getAvailableVoices());
-        }
-        return voices;
+        return getAllVoicesSorted(localeProvider.getLocale());
+    }
+
+    private Set<Voice> getAllVoicesSorted(Locale locale) {
+        Comparator<Voice> byTTSLabel = (Voice v1, Voice v2) -> {
+            return getTTS(v1).getLabel(locale).compareTo(getTTS(v2).getLabel(locale));
+        };
+        Comparator<Voice> byVoiceLocale = (Voice v1, Voice v2) -> {
+            return v1.getLocale().getDisplayName(locale).compareTo(v2.getLocale().getDisplayName(locale));
+        };
+        return ttsServices.values().stream().map(s -> s.getAvailableVoices()).flatMap(Collection::stream)
+                .sorted(byTTSLabel.thenComparing(byVoiceLocale).thenComparing(Voice::getLabel))
+                .collect(Collectors.toCollection(LinkedHashSet::new));
```

Comment:
We could also return a TreeSet here so the set remains sorted when it is modified.

Reply:
With regards to https://github.com/eclipse/smarthome/issues/5960 I suggest to wrap the returned set into an unmodifiable set. Consumers then must create a copy of the set and can decide how to handle modifications. wdyt?

Reply:
I've implemented your suggestion @htreu! Such changes could API breaking though. It is also something to keep in mind with #5960.